### PR TITLE
JS: js/xss: Sanitize attributes that cannot contain quotes

### PR DESCRIPTION
**Pull request by erik-krogh**
Adds a sanitizer guard that checks if a value contains `"`, and if that value ends up on a HTML attribute. 

I also expanded what is recognized as a HTML attribute string concatenation.  

An annoying thing I encountered there is that `\n` or `\r` is not recognized by the regexp `.*`. ([examples](https://lgtm.com/query/1662194255920082447/)).   
This turns out to be completely in spec, as our RegExps are compiled using `Pattern.compile(reg)` without the [`DOTALL` flag](https://docs.oracle.com/javase/8/docs/api/java/util/regex/Pattern.html#DOTALL).

Gets a TN for CVE-2018-3726. 

TODO: 
 - [x] change note
 - [x] evaluation

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/security/dataflow/IncompleteHtmlAttributeSanitizationCustomizations.qll [hunk 1]
```diff
@@ -51,8 +51,11 @@ module IncompleteHtmlAttributeSanitization {
     string lhs;
 
     HtmlAttributeConcatenation() {
-      lhs = this.getPreviousLeaf().getStringValue().regexpCapture("(.*)=\"[^\"]*", 1) and
-      this.getNextLeaf().getStringValue().regexpMatch(".*\".*")
+      lhs = this.getPreviousLeaf().getStringValue().regexpCapture("((?:[\n\r]|.)*)=\"[^\"]*", 1) and
+      (
+        this.getNextLeaf().getStringValue().regexpMatch(".*\".*") or
+        this.getRoot().getConstantStringParts().regexpMatch("(?:[\n\r]|.)*</.*")
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/security/dataflow/IncompleteHtmlAttributeSanitizationCustomizations.qll:57]: The right way to set the dotall flag is to put `(?s)` in the beginning of the regexp (see `HtmlConcatenationRoot`). But yes, it's a rather surprising default.

Speaking of `HtmlConcatenationRoot` could that be used in the second condition instead?

File: javascript/ql/src/semmle/javascript/security/dataflow/Xss.qll [hunk 2]
```diff
@@ -51,10 +54,30 @@ module Shared {
...
+
+    override predicate sanitizes(boolean outcome, Expr e) {
+      e = this.getBaseString().getEnclosingExpr() and outcome = this.getPolarity().booleanNot()
+    }
+  }
 }
 
 /** Provides classes and predicates for the DOM-based XSS query. */
 module DomBasedXss {
+  // StringReplaceCallSequence
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/security/dataflow/Xss.qll:80]:
```suggestion

```

File: javascript/ql/src/semmle/javascript/security/dataflow/IncompleteHtmlAttributeSanitizationCustomizations.qll (hunk 1 above)
Comment (erik-krogh) [reply javascript/ql/src/semmle/javascript/security/dataflow/IncompleteHtmlAttributeSanitizationCustomizations.qll:57]: > The right way to set the dotall flag is to put (?s) in the beginning of the regexp (see HtmlConcatenationRoot). But yes, it's a rather surprising default.

:+1: 

> Speaking of HtmlConcatenationRoot could that be used in the second condition instead?

I'm not sure what you are suggesting. Something like `this.getNextLeaf+().getStringValue().regexpMatch("(?s).*</.*")`?  
(That would require removing `pragma[inline]` from `getNextLeaf()`).

File: javascript/ql/src/semmle/javascript/security/dataflow/IncompleteHtmlAttributeSanitizationCustomizations.qll [hunk 3]
```diff
@@ -51,10 +51,10 @@ module IncompleteHtmlAttributeSanitization {
     string lhs;
 
     HtmlAttributeConcatenation() {
-      lhs = this.getPreviousLeaf().getStringValue().regexpCapture("((?:[\n\r]|.)*)=\"[^\"]*", 1) and
+      lhs = this.getPreviousLeaf().getStringValue().regexpCapture("(?s)(.*)=\"[^\"]*", 1) and
       (
         this.getNextLeaf().getStringValue().regexpMatch(".*\".*") or
-        this.getRoot().getConstantStringParts().regexpMatch("(?:[\n\r]|.)*</.*")
+        this.getRoot().getConstantStringParts().regexpMatch("(?s).*</.*")
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/security/dataflow/IncompleteHtmlAttributeSanitizationCustomizations.qll:57]: This is what I meant:
```suggestion
        this.getRoot() instanceof HtmlConcatenationRoot
```
or equivalently
```suggestion
        this instanceof HtmlConcatenationLeaf
```

**Comment by erik-krogh** [comment]
[Performance looks fine](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-erik-krogh.northeurope.cloudapp.azure.com_1590515938611).

**Comment by asgerf** [comment]
Thanks. LGTM when the conflict is resolved.

**Pull request merged**